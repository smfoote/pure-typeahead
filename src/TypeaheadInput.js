import React, {Component} from 'react'

const noop = () =>{};

export default class TypeaheadInput extends Component {
  static defaultProps = {
    onKeyDown: noop,
    onFocus: noop,
    onKeyUp: noop,
  }

  onChange(evt) {
    this.props.onChange(evt);
  }

  onKeyDown(evt) {
    const { key } = evt;
    switch(key) {
      case 'ArrowDown':
      case 'ArrowUp':
        evt.preventDefault();
        this.props.arrowKeyPressed(key);
        break;
      case 'Enter':
        this.props.enterKeyPressed(evt);
        break;
      case 'Escape':
        this.props.escapeKeyPressed();
        break;
    }
    this.props.onKeyDown(evt);
  }

  focus() {
    this.inputRef.focus();
  }

  render() {
    const { children } = this.props;
    return (
      <input
        ref={input => this.inputRef = input}
        id={this.props.id}
        className={this.props.className}
        type={this.props.type || 'text'}
        placeholder={this.props.placeholder}
        value={this.props.value}
        autoComplete="off"
        onChange={(evt) => this.onChange(evt)}
        onKeyDown={(evt) => this.onKeyDown(evt)}
        onKeyUp={this.props.onKeyUp}
        onFocus={this.props.onFocus}
      />)
  }
}
