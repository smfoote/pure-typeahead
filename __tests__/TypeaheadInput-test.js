import React from 'react';
import Enzyme, { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

import TypeaheadInput from '../src/TypeaheadInput';
import { evtMock } from '../testUtils';

Enzyme.configure({ adapter: new Adapter() });

const shallowSetup = () => {
  const props = {
    value: 'hi',
    onChange: jest.fn(),
    arrowKeyPressed: jest.fn(),
    enterKeyPressed: jest.fn()
  };
  return {
    wrapper: shallow(<TypeaheadInput {...props}/>),
    props
  };
};

describe('Typeahead', () => {
  it('renders self and subcomponents', () => {
    const { wrapper } = shallowSetup();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('calls onChange when a change event happesn', () => {
    const { props, wrapper } = shallowSetup();
    wrapper.find('input').simulate('change', {target: {value: 'hello'}});
    expect(props.onChange).toHaveBeenCalled();
  });

  it('updates value when value prop changes', () => {
    const { props, wrapper } = shallowSetup();
    expect(wrapper.find('input').props().value).toBe('hi');
    wrapper.setProps({value: 'hello'});
    expect(wrapper.find('input').props().value).toBe('hello');
  });

  it('calls arrowKeyPressed when ArrowDown key is pressed', () => {
    const { props, wrapper } = shallowSetup();
    wrapper.find('input').simulate('keydown', {...evtMock, key: 'ArrowDown'});
    expect(props.arrowKeyPressed).toHaveBeenCalled();
  });

  it('calls arrowKeyPressed when ArrowUp key is pressed', () => {
    const { props, wrapper } = shallowSetup();
    wrapper.find('input').simulate('keydown', {...evtMock, key: 'ArrowUp'});
    expect(props.arrowKeyPressed).toHaveBeenCalled();
  });

  it('calls enterKeyPressed when Enter key is pressed', () => {
    const { props, wrapper } = shallowSetup();
    wrapper.find('input').simulate('keydown', {...evtMock, key: 'Enter'});
    expect(props.enterKeyPressed).toHaveBeenCalled();
  });

  it('does not call enterKeyPressed or arrowKeyPressed when a letter is pressed', () => {
    const { props, wrapper } = shallowSetup();
    wrapper.find('input').simulate('keydown', {...evtMock, key: 'a'});
    expect(props.enterKeyPressed).toHaveBeenCalledTimes(0);
    expect(props.arrowKeyPressed).toHaveBeenCalledTimes(0);
  });
});
