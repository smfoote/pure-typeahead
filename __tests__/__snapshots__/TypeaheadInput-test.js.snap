// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeaheadInput renders self and subcomponents 1`] = `
<input
  onBlur={[Function]}
  onChange={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  type="text"
  value="hi"
/>
`;
